src/app/edit-movie-detail/edit-movie-detail.component.ts

```diff
@@ -14,61 +12,34 @@ export class EditMovieDetailComponent implements OnInit {
 
   @Input()
   movieId: string;
-  /*
-  movieForm: FormGroup;
-    id: '';
-    title: '';
-    posterUrl: string = '';
-    genre: string = '';
-    rating: number = null;  */
-
-    movieForm: FormGroup
 
-  constructor(private router: Router, private route: ActivatedRoute, 
-              private movieService: MovieService, private formBuilder: FormBuilder,
-              private activeModal: NgbActiveModal) {
+  form: FormGroup;
 
-                this.movieForm = this.formBuilder.group({
-                  title: [''],
-                  posterUrl: [''],
-                  genre: [''],
-                  rating:['']
-                  })
-              }
+  constructor(
+    private movieService: MovieService,
+    private formBuilder: FormBuilder,
+    private activeModal: NgbActiveModal
+  ) {
 
-  ngOnInit(): void {
-    console.log(this.movieId);
-    this.getMovieById(this.movieId);
-    this.movieForm = this.formBuilder.group({
-      id: [null, Validators.required],
-      title: [null, Validators.required],
-      posterUrl: [null, Validators.required],
-      genre: [null, Validators.required],
-      rating:[null, Validators.required]
+    this.form = this.formBuilder.group({
+      title: ['', [Validators.required]],
+      posterUrl: ['', [Validators.required]],
+      genre: ['', [Validators.required]]
     });
   }
 
-  getMovieById(id: any){
-    this.movieService.getMovieById(id).subscribe((data: any)=>{
-      this.movieForm.setValue({
-        id: data._id,
-        title: data.title,
-        genre: data.genre,
-        rating: data.rating,
-        posterUrl: data.posterUrl
-      });
-    });
+  ngOnInit(): void {
+    this.populatFormWithMovieData(this.movieId);
   }
 
-  onFormSubmit(){
-    this.movieService.updateMovie(this.movieId, this.movieForm.value)
-    .subscribe((res: any) => {
-      this.activeModal.close();
-      location.reload()
-      /*this.movieService.getMovies().subscribe(movies =>{
-        this.movieService.movieList = movies;
-      });*/
-    });
+  async populatFormWithMovieData(id: string) {
+    const movie = await this.movieService.getById(id);
+    this.form.patchValue(movie);
```

Comment:
What is patchValue function about?

Reply:
The patchValue will try to update controls in the form according to the control's name. For example, if I had a form:

`const form = this.formBuilder.group({
    a: [''],
    b: [''],
    c: ['']
});`

and I have an object that I want to populate the form with:

`const data = {
    a: 'hello',
    b: 'world'
};`

Running the following code will merge the values in the `data` object as the form group's control values, changing the values of controls `a` and `b` since those are the only keys available in `data`.

`form.patchValue(data)`

The code above will not trigger an error despite not providing a value for `c`.